test/test_replace.py

```diff
@@ -0,0 +1,171 @@
+@pytest.mark.parametrize(
+    ("dest_config", "src_config", "changes_dict"),
+    [
+        (Level1(name="level1_the_potato"), Level1(), {"name": "level1_the_potato"}),
+        (Level2(name="level2_bar"), Level2(), {"name": "level2_bar"}),
+        (
+            Level2(name="level2_bar", prev=Level1(name="level1_good")),
+            Level2(),
+            {"name": "level2_bar", "prev": {"name": "level1_good"}},
+        ),
+        (
+            Level2(name="level2_bar", prev=Level1(name="level1_good")),
+            Level2(),
+            {"name": "level2_bar", "prev.name": "level1_good"},
+        ),
+        (
+            Level3(
+                name="level3_greatest",
+                prev=Level2(name="level2_greater", prev=Level1(name="level1_greate")),
+            ),
+            Level3(),
+            {
+                "name": "level3_greatest",
+                "prev": {"name": "level2_greater", "prev": {"name": "level1_greate"}},
+            },
+        ),
+        (
+            Level3(
+                name="level3_greatest",
+                prev=Level2(name="level2_greater", prev=Level1(name="level1_greate")),
+            ),
+            Level3(),
+            {
+                "name": "level3_greatest",
+                "prev.name": "level2_greater",
+                "prev.prev.name": "level1_greate",
+            },
+        ),
+    ],
+)
+def test_replace_nested_dataclasses(dest_config: object, src_config: object, changes_dict: dict):
+    config_replaced = replace(src_config, changes_dict)
+    assert config_replaced == dest_config
+
+
+@pytest.mark.parametrize(
+    ("dest_config", "src_config", "changes_dict"),
+    [
+        (InnerPostInit(in_arg=2.0), InnerPostInit(), {"in_arg": 2.0}),
+        pytest.param(
+            OuterPostInit(out_arg=2, inner=(InnerPostInit(3.0, for_outer_post="bar"))),
+            OuterPostInit(),
+            {"out_arg": 2, "inner": {"in_arg": 3.0, "for_outer_post": "bar"}},
+            marks=pytest.mark.xfail(reason='issue_nested_dataclass')
+        ),
+        pytest.param(
+            OuterPostInit(out_arg=2, inner=(InnerPostInit(3.0, for_outer_post="bar"))),
+            OuterPostInit(),
+            {"out_arg": 2, "inner.in_arg": 3.0, "inner.for_outer_post": "bar"},
+            marks=pytest.mark.xfail(reason='issue_nested_dataclass')
+        ),
+    ],
+)
+def test_replace_post_init(dest_config: object, src_config: object, changes_dict: dict):
+    config_replaced = replace(src_config, changes_dict)
+    assert config_replaced == dest_config
+
+
+@pytest.mark.xfail(reason='issue_nested_dataclass')
+def test_replace_mixed_flatten_and_nested_dict():
+    dest_config = Level3(name="PhD", prev=Level2(name="Master", prev=Level1(name="Undergrad")))
+    src_config = Level3()
+
+    replaced_config1 = replace(
+        src_config, {"prev": {"name": "Master", "prev": {"name": "Undergrad"}}}, name="PhD"
+    )
+    assert replaced_config1 == dest_config
```

Comment:
Hmm, I'd be inclined to say that we shouldn't allow this kind of mix. What do you think?

Reply:
That makes sense to me.